doc-surrealdb_versioned_docs/version-2.x/reference-guide/security-best-practices.mdx

```diff
@@ -0,0 +1,323 @@
+
+**Example**
+
+```bash
+# Allow SurrealDB to generate UUIDv4 and call any crypto function except MD5.
+surreal start --allow-funcs "crypto","rand::uuid::v4" --deny-funcs "crypto::md5"
+```
+
+Such a practice becomes specially important in the case of the network capability, which allows SurrealDB to perform network requests such as those required by the [`http::*` functions](https://surrealdb.com/docs/surrealdb/2.x/surrealql/functions/database/http). Allowing untrusted users to perform network requests from your SurrealDB instance can allow them access to its local network or services that specifically allow network access from that instance.
+
+**Example**
+
+```bash
+# Allow SurrealDB to perform HTTP GET requests over to a specific API over TLS.
+surreal start --allow-funcs "http::get" --allow-net "api.example.com:443"
+```
+
+## Passwords
+
+If you require storing passwords for your users, do not rely on table or field permissions to keep them private. In the event that your application or database is compromised, these passwords would become known by the attacker. Instead, use the [password hashing functions](https://surrealdb.com/docs/surrealdb/2.x/surrealql/functions/database/crypto) provided by SurreaDB such  as `crypto::argon2::*`, `crypto::bcrypt::*`, `crypto::pbkdf2::*` and  `crypto::scrypt::*` . These functions ensure that irreversible cryptographic hashes are stored instead of the original passwords, so that the passwords from your users remain safe even in the event of a compromise.
+
+Do not use other cryptographic hash functions (e.g. `crypto::md5`, `crypto::sha1` , `crypto::sha512`) for hashing passwords, even if you do use an additional salt. These functions are designed to be efficient in computing, which will benefit an attacker that sets out to crack any hashes that they may have obtained from the compromise of your application. Hash functions intended for password hashing already incorporate a salt as well as other mechanisms to prevent hash cracking by making the computation of such hashes less efficient. This mitigates password cracking at scale at the small cost of adding a few milliseconds delay while checking credentials for legitimate users.
+
+Even it you only store password hashes, it is a good practice to additionally use field permissions to prevent unauthorised access to the password hashes, which could allow an attacker to perform inefficient but potentially effective attacks such as testing candidate passwords against a specific hash. For even better security, you may store passwords in a separate table and use table permissions to disallow all access to that table. Due to their internal implementation, table permissions provide additional security compared to field permissions.
+
```

Comment:
```suggestion
compromised, these passwords will become known by the attacker. 

Instead, use the [password hashing functions](/docs/surrealdb/2.x/surrealql/functions/database/crypto) provided by SurreaDB, such as `crypto::argon2::*`, `crypto::bcrypt::*`, `crypto::pbkdf2::*` and `crypto::scrypt::*`. These functions ensure that irreversible cryptographic hashes are stored instead of the original passwords so that your users' passwords remain safe even in the event of a compromise.


:::note
__Note:__ Do not use other cryptographic hash functions (e.g. `crypto::md5`, `crypto::sha1` , `crypto::sha512`) for hashing passwords, even if you use an additional salt.
::: 

These functions are designed to be efficient in computing, which will benefit an attacker who sets out to crack any hashes they may have obtained from the compromise of your application. Hash functions intended for password hashing already incorporate a salt and other mechanisms to prevent hash cracking by making the computation of such hashes less efficient. This mitigates password cracking at scale at the small cost of adding a few milliseconds delay while checking credentials for legitimate users.

Even if you only store password hashes, it is a good practice to additionally use field permissions to prevent unauthorized access to the password hashes, which could allow an attacker to perform inefficient but potentially effective attacks such as testing candidate passwords against a specific hash. For even better security, you may store passwords in a separate table and use table permissions to disallow all access to that table. Due to their internal implementation, table permissions provide additional security compared to field permissions.
```

Reply:
The "These functions are designed to be efficient in computing, which will benefit an attacker who sets out to crack any hashes they may have obtained from the compromise of your application." part of the third paragraph refers to the functions that you SHOULD NOT use. It should be inside of the warning if we do want to use that format.

Reply:
I have addressed this in [d56a22c](https://github.com/surrealdb/docs.surrealdb.com/pull/651/commits/d56a22c380c2d35275ec64ebc264ac3c1f21e194). I think the warning is a nice touch but I believe is unnecessary given that this whole guide is essentially a list of warnings. This prevents splitting the paragraph that relates to the content in the warning or having a warning that is too long.

Reply:
That's a fair point. The note would have been bloated anyways